### PR TITLE
Add gauge LM system

### PR DESCRIPTION
**Pull request by chuddster**
Distributes funds to vaults based on sRBN voting weights

- add GaugeController
- add ERC20CompetitiveRewardModule (weighted vault deposit LM)

[crv docs](https://curve.readthedocs.io/dao-gauges.html#implementation-details) for reference

### CODE REVIEW

File: contracts/rbn-staking/GaugeController.vy [hunk 1]
```diff
@@ -551,6 +556,23 @@ def vote_for_gauge_weights(_gauge_addr: address, _user_weight: uint256):
 
     log VoteForGauge(block.timestamp, msg.sender, _gauge_addr, _user_weight)
 
+@external
```
Comment (chuddster) [inline contracts/rbn-staking/GaugeController.vy:559]: this is the method that distributes to all of the gysr staking contracts based on weight voted on. gauge controller already takes care of voting logic based on sRBN (veCRV) balance

File: contracts/rbn-staking/GaugeController.vy [hunk 2]
```diff
@@ -558,16 +558,19 @@ def vote_for_gauge_weights(_gauge_addr: address, _user_weight: uint256):
     log VoteForGauge(block.timestamp, msg.sender, _gauge_addr, _user_weight)
 
 @external
-def disperse_funds(_total_funding: uint256):
+def disperse_funds(_total_funding: uint256, _duration: uint256 = WEEK):
```
Comment (chuddster) [inline contracts/rbn-staking/GaugeController.vy:561]: people can change their weights for a particular gauge once [every 10 days](https://github.com/ribbon-finance/governance/blob/bc4b0867b00e765b47daa2d8b34b0440070e14c1/contracts/rbn-staking/GaugeController.vy#L16) in the crv implementation

File: test/tvl-staking/GaugeController.ts [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+import { addSnapshotBeforeRestoreAfterEach } from "../common";
+import {
+  ONE_WEEK,
+  ONE_YEAR,
+  DEFAULT_DECIMALS,
+  ZERO_ADDRESS,
+  DEAD_ADDRESS,
+} from "../utils/constants";
+chai.use(solidity);
+
```
Comment (chuddster) [inline test/tvl-staking/GaugeController.ts:23]: only added testing on what I changed (disperse_funds).

crv team [testing](https://github.com/curvefi/curve-dao-contracts/tree/master/tests/integration/GaugeController) of gauge controller

File: contracts/rbn-staking/GaugeController.vy [hunk 4]
```diff
@@ -0,0 +1,635 @@
...
+    # For all the gauges, fund each one based on respective weight
+    for gauge in self.gauges:
+      # weight * total_funding / 1e18.
+      # if weight = 1e18, weight = 1.0
+      gauge_funding_amt: uint256 = self._gauge_relative_weight(gauge, block.timestamp) * _total_funding / MULTIPLIER
+      if(gauge_funding_amt == 0):
+        continue
+      ERC20(rbn).approve(gauge, gauge_funding_amt)
+      VaultStaking(gauge).fund(gauge_funding_amt, _duration)
+      log DisperseFunds(gauge, gauge_funding_amt)
```
Comment (kenchangh) [inline contracts/rbn-staking/GaugeController.vy:591]: may be too gas inefficient to log on every gauge, better for us just to look that up from the contract on the UI

File: contracts/rbn-staking/GaugeController.vy (hunk 2 above)
Comment (kenchangh) [reply contracts/rbn-staking/GaugeController.vy:567]: 10 days sounds fine

File: contracts/rbn-staking/GaugeController.vy (hunk 4 above)
Comment (chuddster) [reply contracts/rbn-staking/GaugeController.vy]: agree. can always look for [fund event](https://github.com/ribbon-finance/gysr-contracts/blob/c89bb54530c15040ca63a39f394c0ad838b2e6e1/contracts/ERC20BaseRewardModule.sol#L163) which is whenever we fund()

**Pull request closed without merging**